fplll/hlll.h

```diff
@@ -0,0 +1,175 @@
+/*
+   This file is part of fplll. fplll is free software: you
+   can redistribute it and/or modify it under the terms of the GNU Lesser
+   General Public License as published by the Free Software Foundation,
+   either version 2.1 of the License, or (at your option) any later version.
+
+   fplll is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+   GNU Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public License
+   along with fplll. If not, see <http://www.gnu.org/licenses/>. */
+
+#ifndef FPLLL_HLLL_H
+#define FPLLL_HLLL_H
+
+#include "householder.h"
+#include <cmath>
+
+FPLLL_BEGIN_NAMESPACE
+
+template <class ZT, class FT> class HLLLReduction
+{
+public:
+  /**
+   * Constructor.
+   * The precision of FT must be defined before creating an instance of the
+   * class and must remain the same until the object is destroyed (or no longer
+   * needed).
+   */
+  // TODO: what is c?
+  HLLLReduction(MatHouseholder<ZT, FT> &arg_m, double delta, double eta, double theta, double c,
+                int flags)
+      : m(arg_m)
+  {
+    this->delta = delta;
+    this->eta   = eta;
+    this->theta = theta;
+    this->c     = c;
+    sr          = pow(2.0, -(double)m.get_d() * c);
+    verbose     = flags & LLL_VERBOSE;
+    dR.resize(m.get_d());
+    eR.resize(m.get_d());
+    status = -1;
+  }
+
+  /**
+    @brief Househorder inside LLL reduction.
+    */
+  bool hlll();
+
+  // Get the status of the computation
+  inline int get_status() { return status; }
+private:
+  // Paramters to (delta, eta, theta) hlll-reduce the basis b in m.
+  FT delta, eta, theta;
+  MatHouseholder<ZT, FT> &m;
+
+  // Arbitraty c > 0
+  FT c;
+  // Multiplicative coefficient used to check if a vector is size-reduced or not.
+  FT sr;
+  // Verbose mode.
+  bool verbose;
+
+  // Temporary variables
+  FT ftmp0, ftmp1, ftmp2;
+  long expo0, expo1, expo2;
+
+  int status;
+
+  /**
+     @brief Size reduction.
+
+     Perform size reduction of b[kappa]. Reduce b[kappa] with
+     b[size_reduction_start..size_reduction_end-1].
+
+     @param kappa index of the vector
+  */
+  void size_reduction(int kappa, int size_reduction_end, int size_reduction_start = 0);
+
+  /**
+   * In verbose mode, print informations to reproduce the computation (parameters, enable features)
+   */
+  inline void print_params();
+
+  // Precompute dR[k] * 2^(2*row_expo[k]) = delta_ * R(k, k)^2
+  vector<FT> dR;
+
+  // Compute dR[k]
+  inline void compute_dR(int k);
+
+  // Set the status of the computation and print message if verbose
+  inline bool set_status(int new_status);
+
+  // Precompute eR[k] * 2^row_expo[k] = eta * R(k, k)
+  vector<FT> eR;
+
+  // Compute dR[k]
```

Comment:
Wrong comment

Reply:
Will be fixed with commit 9de2f4a41332aa9500a0f9b471953d084061dc24.